{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "functions_installation_and_tutorial.ipynb",
      "provenance": [],
      "collapsed_sections": [],
      "toc_visible": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "d5P0abPQjjI_"
      },
      "source": [
        "# ExTaxsI Python functions"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "6QYWzOpWjqaz"
      },
      "source": [
        "ExTaxsI functions can be used separately from the ExTaxsI tool. \\\n",
        "In general, there are 6 main functions for the following tasks: \n",
        "\n",
        "\n",
        "*   load_configurations() - **mandatory before using ExTaxsi functions** - setting NCBI API key to allow data query\n",
        "*   db_creation() - create FASTA, accession with taxonomy or enriched database\n",
        "*   taxonomyID_converter() - convert NCBI taxid into organism taxonomy and viceversa\n",
        "*   sunburst_plot() - create sunburst plot\n",
        "*   scatterplot() - create scatterplot\n",
        "*   worldmap_plot() - create worldmap plot\n",
        "\n",
        "Here we report the installation via conda and pip. Furher, tutorial for configuration and usage is provided. Input files used are available in the **examples** directory. Use help() functions to see ExTaxsI function details. In this notebook, all the descriptions of the main functions are reported.\n",
        "\n",
        "\\\\\n",
        "For any doubt, please write to g.agostinetto@campus.unimib.it"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "iiOOavoLjqwx"
      },
      "source": [
        "## Conda installation"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "RVB9pFQYlAoS"
      },
      "source": [
        "Source: https://github.com/qLSLab/ExTaxsI\n",
        "\n",
        "Requirements:\n",
        "\n",
        "\n",
        "*   Python 3 (https://www.python.org/)\n",
        "*   Conda or Miniconda (https://conda.io/projects/conda/en/latest/user-guide/install/index.html)\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "jcBQ3xDyjbmX"
      },
      "source": [
        "# environment creation\n",
        "\n",
        "conda create --name extaxsi -c qlslab -c conda-forge -c etetoolkit\n",
        "\n",
        "# activate\n",
        "\n",
        "conda activate extaxsi\n",
        "\n",
        "# deactivate\n",
        "\n",
        "conda deactivate extaxsi"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "xYsESN7Ljt-R"
      },
      "source": [
        "## PyPI installation"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "HqINq6zgl5p4"
      },
      "source": [
        "Source: https://pypi.org/project/extaxsi/"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "B6mIJKublwpe"
      },
      "source": [
        "# installation via pip \n",
        "\n",
        "pip install extaxsi"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "2xvP1k8yl8_R"
      },
      "source": [
        "## Configuration and usage"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "FTPkyzLqqruC"
      },
      "source": [
        "Configuration step is **mandatory** before using ExTaxsI. \\\n",
        "Please, see https://ncbiinsights.ncbi.nlm.nih.gov/2017/11/02/new-api-keys-for-the-e-utilities/ to create a NCBI account and obtain you API key, then insert your e-mail and API key when running load_configurations()."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "_zu731gm-XQZ"
      },
      "source": [
        "# import functions\n",
        "\n",
        "import extaxsi"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "1yjVcastl-Px"
      },
      "source": [
        "# function description\n",
        "\n",
        "help(load_configurations)"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "02gp91Spqjp8"
      },
      "source": [
        "    Load configurations necessary to NCBI and update taxonomy database.\n",
        "    Keyword arguments:\n",
        "    entrez_email (str) -- email registered on NCBI.\n",
        "    api_key (str) -- api key generated by NCBI.\n",
        "    taxa_database_update (str) -- dowload or update taxonomy database (default 'yes')."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "hKWAEALzqkrK"
      },
      "source": [
        "# taxa_database_update -- download or update taxonomy database (default 'yes').\n",
        "# if you configure the first time, put 'yes' or do not specify\n",
        "\n",
        "load_configurations(\"user@email.com\",\"api_key_here\", taxa_database_update = 'no')"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "oiaApMymow7i"
      },
      "source": [
        "### Database creation"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Ty3DAMrxpX07"
      },
      "source": [
        "help(db_creation)"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "2kJJ-LJWpgis"
      },
      "source": [
        "    Create FASTA, accession with taxonomy or enriched database.\n",
        "    Keyword arguments:\n",
        "    text_search (str) -- manual input from the user, ex. txid8832 (default None).\n",
        "    file_search (str) -- file path from the user (default None).\n",
        "    input_file_type (str) -- 'T','A' or 'O' if the input file is a list of Taxid,\n",
        "                        accession or organism (default None).\n",
        "    additional_query (str) -- add string if you want to restrict your output to a specific\n",
        "                        element, ex. 'COI' (default 0).\n",
        "    fasta_output -- create fasta database (default False).\n",
        "    accession_taxonomy_output -- create file with accession and 6 rank taxonomy (default False).\n",
        "    marker_output -- create gene marker output (default False).\n",
        "    top10_plot -- create top 10 plot of marker output (default False).\n",
        "    enrich_output -- create enriched file, i.e. accession with coordinates or country\n",
        "                     location, gene information and organism (default False).\n",
        "    overwrite_file -- if file already exists ovewrite (default True)."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "4z5RCLm-mjj-"
      },
      "source": [
        "# example 1\n",
        "# input text_search txid8832 = Aix galericulata\n",
        "\n",
        "db_creation(text_search='txid8832',\n",
        "            accession_taxonomy_output=True,\n",
        "            fasta_output=False,\n",
        "            marker_output=True,\n",
        "            top10_plot=True,\n",
        "            enrich_output=True)"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "0kV2cfLx625-"
      },
      "source": [
        "# example 2\n",
        "# input file_search = accession list avaliable in example directory \n",
        "# (https://github.com/qLSLab/ExTaxsI/tree/master/examples)\n",
        "\n",
        "db_creation(file_search='example/A_accession_list_example.tsv',\n",
        "            input_file_type='A',\n",
        "            accession_taxonomy_output=True,\n",
        "            fasta_output=True,\n",
        "            marker_output=True,\n",
        "            top10_plot=True)"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "pWp2bvXKoz6l"
      },
      "source": [
        "### Taxonomy ID converter"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "bsHZhXltpUPk"
      },
      "source": [
        "help(taxonomyID_converter)"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "KJn7bYX5phQO"
      },
      "source": [
        "    Convert taxid into organism taxonomy and viceversa.\n",
        "    Keyword arguments:\n",
        "    text_search (str) -- manual search from the user (default none).\n",
        "    file_search (str) -- file search from the user, see examples on documentations (default None).\n",
        "    input_type (str) -- 'O' organism search, 'T' taxID search (default 'O').\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "mlMSvfOCnzlM"
      },
      "source": [
        "taxonomyID_converter(file_search = 'example/O_organism_list_example.tsv', input_type = 'O')"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ldJKxix6o28e"
      },
      "source": [
        "### Sunburst visualization"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "wRXU71HlpaOY"
      },
      "source": [
        "help(sunburst_plot)"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "fF3oapjEo5qL"
      },
      "source": [
        "    Create sunburst plot.\n",
        "    Keywords arguments:\n",
        "    arg[0] (str) -- path to file, accession with taxonomy output if created with db_creation.\n",
        "    arg[1] (str) -- title of the plot.\n",
        "    filter_value (int) -- enter the minimum number of accession per organism (default 0)."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "TWht-G0kn4f9"
      },
      "source": [
        "sunburst_plot(\"download/txid8832_taxonomy.tsv\", \"example\")"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "QunGnitSo6RL"
      },
      "source": [
        "### Scatter plot visualization"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "MEW7Clw7pb5-"
      },
      "source": [
        "help(scatterplot)"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "soY64MOPpihI"
      },
      "source": [
        "    Create scatterplot\n",
        "    Args:\n",
        "        accession_taxonomy_output ([str]): path to file, accession with taxonomy output if created with db_creation.\n",
        "        title_graph ([str]): title of the plot\n",
        "        filter_value (int, optional): enter the minimum number of accession per organism. Defaults to 0."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "wl8FrReEn-Dj"
      },
      "source": [
        "scatterplot(\"download/txid8832_taxonomy.tsv\", \"example\")"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "jAPsLGMVo9m7"
      },
      "source": [
        "### World map plot visualization"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "tjzVSFEspdPk"
      },
      "source": [
        "help(worldmap_plot)"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "2o8xJBJxpjcT"
      },
      "source": [
        "    Create worldmap plot.\n",
        "    Keywords arguments:\n",
        "    enrich_output -- path to file, enriched output if created with db_creation.\n",
        "    title_map -- title of the plot.\n",
        "    Returns:\n",
        "    plot_world_map -- interactive world map of features distributions."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "OVECXjRfoBS_"
      },
      "source": [
        "worldmap_plot(\"download/txid8832_enriched.tsv\",'example_worldmap')"
      ],
      "execution_count": null,
      "outputs": []
    }
  ]
}